implement recursive_shared_mutex

Greg-Griffith opened the pull request:
__NOTE__ _This PR depends itself does not depend on #1553 because it does not replace any of the mutexs currently used in the code with recursive shared mutexs. However, a recursive shared mutex does require that the threads that call for a lock are std::threads which is a not true of all threads in the code base until #1553 is merged_

__Overview__
This is a recursive shared mutex with the following behavior implemented:
- This mutex has two levels of access, shared and exclusive. Multiple threads can own this mutex in shared mode but only one can own it in exclusive mode.
- A thread is considered to have ownership when it successfully calls either lock or try_lock (for either level of access).
- A thread may recursively call lock for ownership and must call a matching number of unlock calls to end ownership.
- A thread MAY call for shared ownership if it already has exclusive ownership. Internally the mutex will treat this as an additional lock for exclusive ownership (separate counter). When a thread has exclusive ownership and calls to unlock shared ownership by calling unlock_shared this will decrement the lock count on exclusive ownership (on the same separate counter).
- There is internal tracking of how many times a thread locked for shared ownership. A thread can not unlock more times than it locked. Trying to do so will cause an assertion as this is a critical error somewhere in the locking logic. 
- A thread may obtain exclusive ownership if no threads excluding itself have shared ownership by calling try_promotion(). Doing so while other threads have shared ownership will block until all other threads have released their shared ownership. Promoting ownership in this way will "jump the line" of other threads that waiting for exclusive ownership and will cause the thread with shared ownership to become the next thread to obtain exclusive ownership. To avoid deadlocks only one thread may attempt this ownership promotion at a time. If a thread has already done this and is currently waiting for promotion and a different thread tries to request promotion the try_promotion() call will return false.

__Additional BCH specific implementation behavioral expectations__
- If a thread has exclusive ownership and checks if it has shared ownership we should should return true.


__NOTES__
- We use try_promotion() for promotions instead of lock() for two reasons:
    - We want to be able to signal that we did not get the promotion. try_promotion() returns a boolean while lock() doesnt return anything.
    - It is generally discouraged to have a lot of threads potentially calling and waiting for promotions because this creates a sort of race condition where the edited data set will be the same for the first thread to get promoted but all following threads arent guarenteed to be editing the same data that was observed during shared ownership
- if the thread that has exclusive ownership got that ownership via promotion, another thread can not request a promotion to follow it. this prevents threads that use promotion to continuously "cut the line" for exclusive ownership. 


__TODO__

Greg-Griffith commented:
I have updated the OP of this PR to give information about what is currently implemented and what functionality still needs to be added before this is considered complete. @gandrewstone I would like your feedback on the full set of functionality when you have the time.

Marking this as WIP while i work on the TODO list.

Greg-Griffith committed:
implement recursive_shared_mutex

Greg-Griffith committed:
track how many times each thread locks in shared mode and do not let it unlock more times than it locked.

Greg-Griffith committed:
Let a thread request exclusive while holding shared using lock (blocking, no "line cutting") and auto drop its shared while it waits. restore its shared locks after it unlocks exclusive

Greg-Griffith committed:
change shared_unlock entry assert to std::logic_error for rsm_tests.cpp

Greg-Griffith committed:
add tests for internal upgrades/downgrades between shared and exclusive.

Greg-Griffith committed:
add try_promotion, lets a thread promote from shared to exclusive lock

ptschip commented on src/test/rsm_tests.cpp:
```diff
@@ -0,0 +1,201 @@
+
+    // lock shared and add a number while holding write lock
+    rsm.lock_shared(std::this_thread::get_id());
+    rsm_guarded_vector.push_back(2);
```
Not sure if you covered it , but I didn't see that you did two or more shared locks in a row...the locks I see but just not the shared_locks.  Would also be good I think to do more than just 2 in a row...could do many in a loop...(what are the limits , are there any?)

Greg-Griffith replied:
added test that recursively locks 100 times to address this concern. i recall reading somewhere that there is an unknown internal limit in c++  for mutex locking but it is higher than should ever be hit by proper normal use.

ptschip commented on src/test/rsm_tests.cpp:
```diff
@@ -0,0 +1,201 @@
+
+BOOST_AUTO_TEST_CASE(RsmTest)
+{
+    std::thread first (alpha);
```
Would like to see a more commenting on what the intent of each test is and what the expected outcomes

Greg-Griffith replied:
in an address to all 3 comments, I will reformat the tests file today for clarity about what each test is testing and to make sure I cover any missed scenarios.

ptschip commented on src/recursive_shared_mutex.h:
```diff
@@ -0,0 +1,102 @@
+    void lock_shared(const std::thread::id &locking_thread_id);
+    bool try_lock_shared(const std::thread::id &locking_thread_id);
+    void unlock_shared(const std::thread::id &locking_thread_id);
+};
```
are there tests for all of these...not sure I saw try_lock_shared().

Greg-Griffith replied:
added another test file that tests all the basic functions in a less complex single threaded way

ptschip commented on src/recursive_shared_mutex.h:
```diff
@@ -0,0 +1,102 @@
+
+
+/*
+This mutex has two levels of access, shared and exclusive. Multiple threads can own this mutex in shared mode but only one can own it in exclusive mode.
```
Preserve our Doxygen style commenting.

+/**
 +*  your text...
 +*  your text..
 +*/

Greg-Griffith replied:
fixed. i forgot to add the new file to the formatting list.

Greg-Griffith committed:
add rsm related files to .formatted-files

Greg-Griffith committed:
formatting fixes

Greg-Griffith commented:
putting this back in WIP while i restructure the tests and rework the error system to not use asserts

Greg-Griffith committed:
replace assertions with std::logic errors

Greg-Griffith committed:
increase rsm test coverage

gandrewstone commented on src/recursive_shared_mutex.h:
```diff
@@ -0,0 +1,101 @@
+    recursive_shared_mutex(const recursive_shared_mutex &) = delete;
+    recursive_shared_mutex &operator=(const recursive_shared_mutex &) = delete;
+
+    void lock(const std::thread::id &locking_thread_id);
```
please add doxygen-like docstring comments for all these APIs

Greg-Griffith replied:
will add documentation for all of the methods

gandrewstone commented on src/recursive_shared_mutex.h:
```diff
@@ -0,0 +1,101 @@
+    recursive_shared_mutex &operator=(const recursive_shared_mutex &) = delete;
+
+    void lock(const std::thread::id &locking_thread_id);
+    bool try_promotion(const std::thread::id &locking_thread_id);
```
as an example of what to document:  What is the state relation between the APIs for example, I think that: state {lock()} == state { lock_shared(); assert(try_promotion()); }

also, document when the APIs are usable, and what happens when called from the wrong state.  For example, calling try_promotion when state == unlock or state == lock.

gandrewstone commented on src/recursive_shared_mutex.h:
```diff
@@ -0,0 +1,101 @@
+    void unlock(const std::thread::id &locking_thread_id);
+    void lock_shared(const std::thread::id &locking_thread_id);
+    bool try_lock_shared(const std::thread::id &locking_thread_id);
+    void unlock_shared(const std::thread::id &locking_thread_id);
```
Why pass the thread id?  Shouldn't we access it from the current thread?  Even if there's a good reason that thread A might call these APIs on behalf of thread B, it seems like the most common use will be the "this" thread, so we should have a default parameter that supplies it.

Greg-Griffith replied:
It was useful early on for debugging but serves little to no purpose now that it is done. i will remove this and just grab this_thread's id when needed

gandrewstone commented on src/recursive_shared_mutex.h:
```diff
@@ -0,0 +1,101 @@
+
+class recursive_shared_mutex
+{
+private:
```
Please make variables protected not private so that test and debugging code can derive from your class and break through the veil.

Greg-Griffith replied:
good idea, will do

gandrewstone commented on src/recursive_shared_mutex.h:
```diff
@@ -0,0 +1,101 @@
+    std::condition_variable _promotion_write_gate;
+
+    // holds a list of owner ids that have shared ownership and the number of times they locked it
+    std::map<std::thread::id, uint64_t> _read_owner_ids;
```
Let's imagine that all the threads used the same count variable.  This means that we couldn't detect (at the moment of the error) the situation where a thread erroneously unlocked too many times since it might not decrement the count below 0 if other threads also have the lock.  

But if we assume a correct program (in terms of matched lock/unlocks), then I think that the semantics of this object would be the same.  And our RAII (scope based) lock objects are a compile-time technique that ensures that threads match their locks and unlocks.

So depending on the performance required it might make sense to do that rather than have a map.  But the map implementation will always be valuable in a "debug" build.  

So let's keep it this way now and think about higher performing choices if needed.

Greg-Griffith replied:
The std::map is needed for promotions. Without the map we can not efficiently know how many of the shared locks the promoting thread. 
It needs to know how many of the shared locks belong to the thread that is promoting so it can move them to the auto locks. this leaves the remaining map with all other threads that need to release shared ownership before exclusive ownership can be obtained via promotion.

gandrewstone replied:
good point

Greg-Griffith committed:
fix issue where threads could grab shared while a thread was trying to promote

Greg-Griffith committed:
change variables from private to protected

Greg-Griffith committed:
remove redundant fuctions

Greg-Griffith committed:
shared ownership can no longer change to exclusive via internal unlocking. The shared onwership now needs to be released before trying to get exclusive ownership or the thread must promote to exclusive ownership through try_promotion

Greg-Griffith commented:
OP has been updated to reflect behavior change in the last few commits. 

TODO: 
- document api calls
- add test for thread "starvation" when attempting to promote (promotion should block additional threads from grabbing shared ownership)

Greg-Griffith commented:
OP has been updated to reflect behavior change in the last few commits. 

TODO: 
- document api calls
- add test for thread "starvation" when attempting to promote (promotion should block additional threads from grabbing shared ownership)
- remove thread id params from external api calls

ptschip commented on src/recursive_shared_mutex.h:
```diff
@@ -0,0 +1,104 @@
+
+
+/*
+ * This mutex has two levels of access, shared and exclusive. Multiple threads can own this mutex in shared mode but
```
almost right for doxygen
At the top it should be

-/*
+/**

Greg-Griffith replied:
fixed. thanks. i thought clang formatter would catch that for me. guess not

Greg-Griffith committed:
public api no longer needs thread id as a param

Greg-Griffith committed:
add comments to all public api methods

Greg-Griffith committed:
simplify promotions and lock counts. add method for end of exclusive

remove _write_promotion_counter and the promotion_read_gate to simplify 
keeping track of locks. to determine if a thread is waiting for a 
promotion we should now just check the _promotion_candidate_id

Greg-Griffith committed:
restructure tests, add starvation tests, fix formatting

Greg-Griffith committed:
replace write_counter==0 checks with end_of_exclusive_ownership()

this prevents a thread with exclusive ownership from not properly
releasing all its shared ownership it aquired while having the exclusive
ownership.

Greg-Griffith committed:
fix formatting

gandrewstone commented on src/recursive_shared_mutex.cpp:
```diff
@@ -0,0 +1,338 @@
+    else
+    {
+        // Wait until we can set the write-entered.
+        _read_gate.wait(_lock, [=] { return end_of_exclusive_ownership(); });
```
I think that there are a few reasons to not use the full capture by value [=] in these lambdas, although the code works.

1. efficiency: you are making copies of unnecessary variables
2. you are copying _lock which might have weird semantics
3. it almost looks like a bug because = captures by value and so a surface reading would imply that _write_promotion_counter should not change. It works because _write_promotion_counter is a member variable and you are really capturing "this" by value and so effectively _write_promotion_counter by reference. But this is quite confusing

Instead why not use [this]?

This comment applies to most or maybe all of your lambdas...

Greg-Griffith replied:
[=] is what is used by std::shared_mutex but you are right about the unnecessary copies and it turns out [this] is slightly more performant.  i have adjusted them to use [this]

Greg-Griffith committed:
change lambdas from [=] to [this] to prevent making copies of unnecessary variables

Greg-Griffith committed:
release exclusive ownership even if all shared locks that were locked while exclusive have not yet been unlocked.
This is to put the mutex in the same state as the intended even though
there is a logic error elsewhere in the program.

Greg-Griffith committed:
no longer auto lock/unlock shared locks on promotion.

Greg-Griffith committed:
fix asymmetry with check_for_write_lock/unlock functions.

Greg-Griffith committed:
wrap assertions and std::logic errors in DEBUG_ASSERTION definitions and update tests to reflect the changes

Greg-Griffith committed:
add copyright, add readme, move rsm files into their own folder to make pulling updates from an upstream repo easier
